CRAMcodecs.tex

```diff
@@ -0,0 +1,3182 @@
+  \State $len \gets $\Call{ReadUint7}{} \Comment{Length of uncompressed O0/O1 data, pre-expansion}
+  \If{$rle\_meta\_len \bitand 1$}
+    \State $rle\_meta \gets $\Call{ReadData}{$\lfloor{}rle\_meta\_len/2\rfloor{}$}
+  \Else
+    \State $comp\_meta\_len \gets $\Call{ReadUint7}{}
+    \State $rle\_meta \gets $\Call{ReadData}{$comp\_meta\_len$}
+    \State $rle\_meta \gets $\Call{RansDecodeNx16\_0}{$rle\_meta\_len/2$, $source = rle\_meta$, $N$}
+    \EndIf
+
+  \Statex
+  \State $n \gets$ \Call{ReadUint8}{$metadata$}
+  \If{$n = 0$}
+    \State $n \gets 256$
+  \EndIf
+  \For{$i \gets 0 \algorithmicto n-1$}
+    \State $s \gets$ \Call{ReadUint8}{$metadata$}
+    \State $L_s \gets 1$
+  \EndFor
+  \Statex
+  \State \Return ($L$, $rle\_meta$, $len$)
+\EndFunction
+\end{algorithmic}
+
+The use of the run length meta-data occurs when expanding the
+uncompressed data, after Order-0 or Order-1 data decompression.
+
+\begin{algorithmic}[1]
+\Statex (Expands data ($in$) using run-length metadata)
+\Function{DecodeRLE}{$in$, $L$, $metadata$, $in\_len$}
+  \State $j \gets 0$
+  \For{$i \gets 0 \algorithmicto in\_len - 1$}
+    \State $sym \gets$ \Call{ReadUint8}{$in$}
+    \If{$L_s$}
+      \State $run \gets$ \Call{ReadUint7}{$metadata$}
+      \For{$k \gets 0 \algorithmicto run$}
+          \State $out_{j+k} \gets s$
+      \EndFor
+      \State $j \gets j + run + 1$
+    \Else
+      \State $out_j \gets s$
+      \State $j \gets j + 1$
+    \EndIf
+  \EndFor
+  \State \Return $out$
+\EndFunction
+\end{algorithmic}
+
+
+\subsection{rANS Nx16 Bit Packing}
+\label{sec:ranspack}
+
+If the alphabet of used symbols in the uncompressed data stream is
+small - no more than 16 - then we can pack multiple symbols together
+to form bytes prior to compression.  This permits 2, 4, 8 and infinite
+(all symbols are the same) numbers per byte.  The distinct
+symbol values do not need to be adjacent as a mapping table $P$
+converts mapped value $x$ to original symbol $P_x$.
+
+The packed format is split into uncompressed meta-data (below) and the compressed packed data.
+
+\begin{table}[h]
+\centering
+% \caption {Meta data format}
+\begin{tabular}{lllp{9cm}}
+\textbf{Bytes} & \textbf{Type} & \textbf{Name} & \textbf{Description}\\
+\hline
+1      & byte   & $nsym$ & Number of distinct symbols\\
+$nsym$ & byte[] & $syms$ & Symbol map \\
+-?     & int7   & $len$  & Length of packed data
+\end{tabular}
+\end{table}
+
+The first meta-data byte holds $nsym$, the number of distinct values, followed
+by $nsym$ bytes to construct the $P$ map.  If $nsym = 1$ then the byte
+stream is a stream of constant values and no bit-packing is done (we
+know every value already).  If $nsym = 2$ then each symbol is 1 bit (8
+per byte), if $2 < nsym \le 4$ symbols are 2 bits each (4 per byte)
+and if $4 < nsym \le 16$ symbols are 4 bits each (2 per byte). It is
+not permitted to have $nsym > 16$ or $nsym = 0$ as bit packing is not
+possible. Bits are unpacked from low to high.
```

Comment:
See https://github.com/samtools/htscodecs/issues/65. htscodecs currently writes `nsym` = 0 when the input has 255 symbols.

Reply:
Thanks.  I've fixed this in https://github.com/samtools/htscodecs/pull/66